tests/integration/credset_test.go

```diff
@@ -0,0 +1,516 @@
+			APIVersion: "porter.sh/v1",
+			Kind:       "CredentialSet",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "porter-test-me-",
+		},
+		Spec: porterv1.CredentialSetSpec{
+			//TODO: get schema version from porter version?
+			SchemaVersion: schemaVersion,
+			Name:          csName,
+		},
+	}
+	return cs
+}
+
+func createTestSecret(ctx context.Context, name, value, namespace string) {
+	Log(fmt.Sprintf("create k8s secret '%s' for credset", name))
+	csSecret := NewTestSecret(name, value)
+	csSecret.ObjectMeta.Namespace = namespace
+	Expect(k8sClient.Create(ctx, csSecret)).Should(Succeed())
+}
+
+func NewTestSecret(name, value string) *corev1.Secret {
+	csSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Type:       corev1.SecretTypeOpaque,
+		StringData: map[string]string{"value": value},
+	}
+	return csSecret
+}
+
+func NewTestInstallation(iName string) *porterv1.Installation {
+	inst := &porterv1.Installation{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "porter.sh/v1",
+			Kind:       "Installation",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "porter-test-me-",
+		},
+		Spec: porterv1.InstallationSpec{
+			SchemaVersion: schemaVersion,
+			Name:          iName,
+			Bundle: porterv1.OCIReferenceParts{
+				Repository: "ghcr.io/bdegeeter/porter-test-me",
+				Version:    "0.3.0",
+			},
+		},
+	}
+	return inst
+}
+
+func newAgentAction(namespace string, name string, cmd []string) *porterv1.AgentAction {
+	return &porterv1.AgentAction{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Spec: porterv1.AgentActionSpec{
+			Args: cmd,
+		},
+	}
+}
+
+func waitForAgentAction(ctx context.Context, aa *porterv1.AgentAction, msg string) error {
+	Log("%s: %s/%s", msg, aa.Namespace, aa.Name)
+	key := client.ObjectKey{Namespace: aa.Namespace, Name: aa.Name}
+	ctx, cancel := context.WithTimeout(ctx, getWaitTimeout())
+	defer cancel()
+	for {
+		select {
+		case <-ctx.Done():
+			Fail(errors.Wrapf(ctx.Err(), "timeout %s", msg).Error())
+		default:
+			err := k8sClient.Get(ctx, key, aa)
+			if err != nil {
+				// There is lag between creating and being able to retrieve, I don't understand why
+				if apierrors.IsNotFound(err) {
+					time.Sleep(time.Second)
+					continue
+				}
+				return err
+			}
+
+			// Check if the latest change has been processed
+			if aa.Generation == aa.Status.ObservedGeneration {
+				if apimeta.IsStatusConditionTrue(aa.Status.Conditions, string(porterv1.ConditionComplete)) {
+					return nil
+				}
+
+				if apimeta.IsStatusConditionTrue(aa.Status.Conditions, string(porterv1.ConditionFailed)) {
+					// Grab some extra info to help with debugging
+					//debugFailedCSCreate(ctx, aa)
+					return errors.New("porter did not run successfully")
+				}
+			}
+
+			time.Sleep(time.Second)
+			continue
+		}
+	}
+
+}
+
+func waitForPorterCS(ctx context.Context, cs *porterv1.CredentialSet, msg string) error {
```

Comment:
I wonder if some of these helper functions can be made more generic with an Object interface instead or any other tricks so that we don't have to duplicate these functions for each type? 🤔 

If you figure that out, it would make a great follow-up PR.

Reply:
agreed, we should follow up to remove this dup